[Model Monitoring] Deploy the default app on the go

jond01 opened the pull request:
Resolves [ML-5865](https://iguazio.atlassian.net/browse/ML-5865).
Move the deployment to the backend in the background.
This enables a smoother and non-blocking experience with the `project.enable_model_monitoring()` call.

The relevant system tests passed.

[ML-5865]: https://iguazio.atlassian.net/browse/ML-5865?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jond01 committed:
Add a method to deploy the histogram drift app

jond01 committed:
Add API POST endpoint

jond01 committed:
Partially backport HTTPMethod from Python 3.11

https://docs.python.org/3/library/http.html#http.HTTPMethod

jond01 committed:
Add the deploy method to HTTPRunDB

jond01 committed:
Use the deploy method in the "enable mm" method

Fix wrong type hints

jond01 committed:
Remove an unsed function

jond01 committed:
Small refactor

jond01 committed:
Workaround for the default hist dd app

jond01 committed:
Don't import `histogram_data_drift`

https://github.com/mlrun/mlrun/actions/runs/8470757532/job/23209199517?pr=5344#step:5:23

jond01 committed:
Inherit from the special enum class with list

jond01 committed:
Refactor `disable_model_monitoring`

jond01 committed:
Delete histogram app in disable_model_monitoring

jond01 committed:
Add mock testing for disable_model_monitoring

davesh0812 commented on server/api/crud/model_monitoring/deployment.py:
```diff
@@ -479,6 +487,56 @@ def _check_if_already_deployed(
+                project=self.project,
+                writer_application_name=mm_constants.MonitoringFunctionNames.WRITER,
+            )
+        ).respond()
```
Please take the same code we using in `project._instantiate_model_monitoring_function()` and copy the relevant part to one function that can be used from both server and mlrun - I think you can put it under `/mlrun/mlrun/model_monitoring/api.py`,  and if you want to take it to next step you can copy 'set_model_monitoring_function()' to this file and when we will be able in the future to add another api of set_mm_function.

davesh0812 commented on mlrun/projects/project.py:
```diff
@@ -2041,22 +2041,14 @@ def enable_model_monitoring(
-            )
-            fn.deploy()
+            result |= db.deploy_histogram_data_drift_app(project=self.name, image=image)
+        return result
```
This dictionary is very confusing and I don't think we need to return it to the user.
You can maybe print it as a debug but no more

davesh0812 commented on mlrun/projects/project.py:
```diff
@@ -2041,22 +2041,14 @@ def enable_model_monitoring(
             image = default_controller_image
         db = mlrun.db.get_run_db(secrets=self._secrets)
-        db.enable_model_monitoring(
+        result = db.enable_model_monitoring(
```
please add option to deploy the histogram_data_drift_app from this api as well, to avoid unnecessary api calls.

jond01 replied:
The API call costs nothing, and I prefer not to add the app to the infrastructure deployment.

For reference, the session is not created for every request:
https://github.com/mlrun/mlrun/blob/65a14413a56c3dbb4ab54df2417c1938142e0e46/mlrun/db/httpdb.py#L250

Edit: done. It's now a single API call.

davesh0812 commented on mlrun/projects/project.py:
```diff
@@ -2041,22 +2041,14 @@ def enable_model_monitoring(
-                image=image,
-            )
-            fn.deploy()
+            result |= db.deploy_histogram_data_drift_app(project=self.name, image=image)
```
You don't think we need also sdk for this?

davesh0812 commented on mlrun/projects/project.py:
```diff
@@ -2012,7 +2012,7 @@ def enable_model_monitoring(
         base_period: int = 10,
         image: str = "mlrun/mlrun",
         deploy_histogram_data_drift_app: bool = True,
```
I remember we agreed on adding an waiting flag with default set to false.
You can use `_wait_for_function_deployment()`.

jond01 replied:
It's out of scope for the current PR, but I agree it can be helpful. Let's address it in another PR.

davesh0812 commented on mlrun/projects/project.py:
```diff
@@ -2081,20 +2073,21 @@ def update_model_monitoring_controller(
+    def disable_model_monitoring(
+        self, *, delete_histogram_data_drift_app: bool = True
+    ) -> None:
+        """
+        Disable model monitoring by deleting the underlying functions infrastructure from MLRun database.
+
+        :param delete_histogram_data_drift_app: Whether to delete the histogram data drift app.
+        """
```
In the end of 1.7.0 this function will delete also the nuclio pods of those function, and without this ability this function is more harm than good, therefore I think for now lets mention it in the docs and maybe we can write also that this function is not valid for now.

davesh0812 reviewed (changes requested):
Nice!
I had few questions and requests :)

jond01 committed:
Comment on the `disable_model_monitoring` method

jond01 committed:
Don't return the function dictionary from the deployment

jond01 committed:
Add deploy_histogram_data_drift_app method to the project

jond01 committed:
Remove unused imports

jond01 committed:
Remove an unused label

jond01 committed:
Use constants instead of magic strings

jond01 committed:
Use project.create_model_monitoring_function

Reduce code duplication

jond01 committed:
Remove duplicated parts from set_model_monitoring_function

This part happens in _instantiate_model_monitoring_function

jond01 committed:
Fix type hints

jond01 committed:
Use mm_consts

jond01 committed:
Export _create_model_monitoring_function_base to a function

jond01 committed:
Update `deploy_histogram_data_drift_app`

jond01 committed:
Add deploy_histogram_data_drift_app param to enable_mm

jond01 committed:
Change the import to avoid circular imports

AttributeError: partially initialized module 'mlrun' has no attribute 'DataItem' (most likely due to a circular import)

jond01 commented on mlrun/projects/project.py:
```diff
@@ -1935,48 +1927,40 @@ def create_model_monitoring_function(
+        requirements: typing.Union[str, list[str], None] = None,
         requirements_file: str = "",
         **application_kwargs,
     ) -> tuple[str, mlrun.runtimes.BaseRuntime, dict]:
+        import mlrun.model_monitoring.api
+
         function_object: RemoteRuntime = None
```
This import is inside the method to avoid a circular import when it's done on the module level

jond01 committed:
Fix `function_obj` -> `func`

davesh0812 commented on mlrun/db/base.py:
```diff
@@ -726,5 +726,11 @@ def enable_model_monitoring(
-        pass
+        deploy_histogram_data_drift_app: bool = True,
+    ) -> None:
+        raise NotImplementedError
```
All the other unimplemented methods in this class are passed not raising anything, I think better to stick with the same behavior.

assaf758 replied:
agree should be consistent, we need to decide if should raise an expception or to pass.

jond01 replied:
Addressed in #5418 as well as the other comments.

davesh0812 commented on mlrun/db/base.py:
```diff
@@ -726,5 +726,11 @@ def enable_model_monitoring(
+    def deploy_histogram_data_drift_app(
+        self, project: str, image: str = "mlrun/mlrun"
+    ) -> None:
+        raise NotImplementedError
```
same here

davesh0812 commented on server/api/crud/model_monitoring/deployment.py:
```diff
@@ -479,6 +488,37 @@ def _check_if_already_deployed(
                 pass
         logger.info(f"Deploying {function_name} function", project=self.project)
 
+    def deploy_histogram_data_drift_app(self, image: str) -> None:
```
Return should be the same as the other deploy functions

davesh0812 reviewed (approved):
LGTM, 
few comments, subject to @assaf758  decision

assaf758 reviewed (approved):
Looks good.
Merging this PR, but @jond01 please address the comments in a follow-up PR.
